### PR TITLE
[FIX] mail: parse eml/msg attachments

### PR DESCRIPTION
**Pull request by len-foss**
When parsing an email with some eml or dmg attachment, the parsing would
generate an attachment with empty content.
Since this content is later encoded in base64, it would crash trying to encode
None.
We simply do not add attachments with empty content.
However this leads to the attachment having for filename the generic
"attachment" (without extension) instead of say "original_message.eml".
This is because the parsing would need to take into account the tree structure
of the message instead of the flattened DFS walk currently done via
message.walk().
Doing so would however necessitate to refactor significantly the mail parsing.

We add a test to check that the parsing does not generate empty content,
since the crash implies that the mail gets stuck on server
and is never fully processed by Odoo.

opw 1814014

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/tests/test_mail_gateway.py [hunk 1]
```diff
@@ -299,6 +299,150 @@
 """
 
 
+MAIL_EML_ATTACHMENT = """Subject: Re: test attac
+From: William <whe@odoo.com>
```
Comment (nim-odoo) [inline addons/mail/tests/test_mail_gateway.py]: I would make the test mail anonymous (no real name, mail, email or IP addresses).

**Comment by nim-odoo** [comment]
@odony I would appreciate your feedback since it touches a sensitive topic (mail).

File: addons/mail/tests/test_mail_gateway.py [hunk 2]
```diff
@@ -366,6 +510,51 @@ def test_message_parse(self):
...
         self.assertEqual(res['attachments'][0][0], 'thetruth.pdf')
 
+    @mute_logger('odoo.addons.mail.models.mail_thread')
+    def test_message_parse_eml(self):
+        """ Test parsing of mail with embedded emails as eml(msg) files.
+        """
+
+        # Tree structure of the message:
+        def display(message_as_tree):
+            return d(message_as_tree, "", "")
```
Comment (tde-banana-odoo) [inline addons/mail/tests/test_mail_gateway.py]: Please avoid nested methods with names that do not mean anything.

File: addons/mail/tests/test_mail_gateway.py [hunk 3]
```diff
@@ -366,6 +510,51 @@ def test_message_parse(self):
...
+            m = ", ".join([bound, content_type, filename, content])
+            pprint_tree += prefix + m + "\n"
+            if isinstance(tree.get_payload(), str):  # a fucking leaf
+                pass
+            else:
+                for son in tree.get_payload():
+                    pprint_tree = d(son, "----" + prefix, pprint_tree)
+            return pprint_tree
+
+        """ display(res) (i.e. MAIL_EML_ATTACHMENT) results in:
```
Comment (tde-banana-odoo) [inline addons/mail/tests/test_mail_gateway.py]: It seems to only assert is totally not linked to the display method. What is its purpose ?

Comment (len-odoo) [reply addons/mail/tests/test_mail_gateway.py]: The purpose of the display is to see the tree structure of the mail.
As explained, the filename is currently not extracted because the parsing does not take into account the structure. 
To get rid of the TODO, you would need to actually take it into account, so this display is to make that explicit.

File: addons/mail/tests/test_mail_gateway.py (hunk 2 above)
Comment (xmo-odoo) [reply addons/mail/tests/test_mail_gateway.py]: Also if all "display" does is call d with two parameters set to a default value, why not just use actual default values on the main function?

Comment (len-odoo) [reply addons/mail/tests/test_mail_gateway.py]: Because technically d is a private method of display, that I didn't nest.
Also this writing style is common because you generally can't avoid it if you want to write tail recursive functions.

Comment (xmo-odoo) [reply addons/mail/tests/test_mail_gateway.py]: I'm reading that `d` is useless. This is a helper function of a test module, it can just sit at the toplevel of the module and take two optional parameters. And a docstring or better naming to explain what it actually does.

File: addons/mail/tests/test_mail_gateway.py (hunk 3 above)
Comment (tde-banana-odoo) [reply addons/mail/tests/test_mail_gateway.py]: My question seems still unanswered. Unless I am mistaken it is not called and therefore unnecessary.

File: addons/mail/models/mail_thread.py [hunk 4]
```diff
@@ -1734,7 +1738,7 @@ def _message_post_process_attachments(self, attachments, attachment_ids, message
                 content = content.encode('utf-8')
             data_attach = {
                 'name': name,
-                'datas': base64.b64encode(content),
+                'datas': base64.b64encode(content or ""),
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py]: If we have nothing to encode shouldn't we avoid creating void attachments instead ?

File: addons/mail/models/mail_thread.py [hunk 5]
```diff
@@ -1441,32 +1441,36 @@ def _message_extract_payload(self, message, save_original=False):
...
-                    html = tools.ustr(part.get_payload(decode=True), encoding, errors='replace')
+                    html = tools.ustr(part_payload, encoding, errors='replace')
                     if not append_content:
                         body = html
                     else:
                         body = tools.append_content_to_html(body, html, plaintext=False)
                 # 4) Anything else -> attachment
                 else:
-                    attachments.append(self._Attachment(filename or 'attachment', part.get_payload(decode=True), {}))
+                    if part_payload:
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py]: You could use an elif here to shorten a bit.

File: addons/mail/tests/test_mail_gateway.py (hunk 3 above)
Comment (len-odoo) [reply addons/mail/tests/test_mail_gateway.py]: Currently the parsing does not get the filename of the attachment because it doesn't take into account the tree structure of the email.
To be able to do it, you would have to use this tree structure. 
This function helps visualize the tree structure to see how things should be done.

Since this bug would require a massive rewrite of the parsing, it cannot be done in stable. 
However if somebody wants to rewrite this correctly, it is a little hint to get it right.

Comment (tde-banana-odoo) [reply addons/mail/tests/test_mail_gateway.py]: Ok. Then I would better have a commit message explaining how it should be done completely, why it cannot be done completely, and remove this code, especially it is used for debug purpose and is not part of the parsing. I think commit message is the right place to explain things. If you think we could improve it in master, feel free to create a task with complete explanations so that we can work on it. Thanks !

Comment (len-odoo) [reply addons/mail/tests/test_mail_gateway.py]: Fine, I now see the errors of my way. I won't bother you with this anymore.
